File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>
using namespace std;

int answer = 0;

void dfs(int index, int sum, int target, vector<int> numbers) {
    if(index == numbers.size()) {
        if(sum == target) {
            answer++;
        }
        return;
    }
    dfs(index+1, sum+numbers[index], target, numbers);
    dfs(index+1, sum-numbers[index], target, numbers);
}

int solution(vector<int> numbers, int target) {
    answer = 0;
    dfs(0, 0, target, numbers);
    return answer;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;
/*
    nubmers의 숫자들을 양수인경우와 음수인경우의 모든 수를 구하고
    합이 target일떄 return;
    0
    알고리즘: DFS, 부분집합
*/
class Solution {
    static int cnt;
    public int solution(int[] numbers, int target) {
        cnt = 0;
        dfs(0, 0, numbers, target);
        
        return cnt;
    }
    static void dfs(int sum, int depth, int[] numbers, int target) {
        if(depth >= numbers.length) {
            if(sum == target) {
                cnt++;
            }
            return;
        }
        
        dfs(sum + numbers[depth], depth+1, numbers, target);
        dfs(sum - numbers[depth], depth+1, numbers, target);
    }
}